test/test_schemes/initialize_constituents.F90

```diff
@@ -0,0 +1,140 @@
+                                                 'cnst_RAINQM', &
+                                                 'cnst_CLDICE', &
+                                                 'cnst_SNOWQM', &
+                                                 'cnst_GRAUQM'/)
+    character(len=11), parameter :: water_species_number_concentrations(5) = &
+                                      (/'cnst_NUMLIQ', &
+                                        'cnst_NUMRAI', &
+                                        'cnst_NUMICE', &
+                                        'cnst_NUMSNO', &
+                                        'cnst_NUMGRA'/)
+
+    errcode = 0
+    errmsg = ''
+    constituent_index = 0
+    found_const_count = 0
+
+    ncdata => initial_file_get_id()
+    ! See how many variables are present on the file
+    ierr = pio_inquire(ncdata, nVariables=num_variables)
+    allocate(constituent_names(num_variables), stat=ierr)
+    if (ierr /= 0) then
+       errcode = 1
+       errmsg = 'Failed to allocate "constituent_names"'
+       return
+    end if
+
+    ! Loop over all variables in the file and add each constituent to the
+    !  dynamic constituent array
+    do var_index = 1, num_variables
+       ierr = pio_inq_varname(ncdata, var_index, variable_name)
+       known_constituent = .false.
+       if (index(variable_name, 'cnst_') > 0) then
+          constituent_index = constituent_index + 1
+          ! Replace with standard name if known, to avoid duplicates
+          if (found_const_count < size(const_std_names)) then
+             do known_const_index = 1, size(const_file_names)
+                if (trim(const_file_names(known_const_index)) == trim(variable_name)) then
+                   constituent_names(constituent_index) = const_std_names(known_const_index)
+                   found_const_count = found_const_count + 1
+                   known_constituent = .true.
+                   exit
+                end if
+             end do
+          end if
+          if (.not. known_constituent) then
+             constituent_names(constituent_index) = variable_name
+          end if
+       end if
+    end do
+
+    allocate(constituents(constituent_index), stat=ierr)
+    if (ierr /= 0) then
+       errcode = 1
+       errmsg = 'Failed to allocate "constituents"'
+       return
+    end if
+
+    do var_index = 1, size(constituents)
+       if (any(water_species_number_concentrations == trim(constituent_names(var_index)))) then
+          call constituents(var_index)%instantiate( &
+             std_name = constituent_names(var_index), &
+             long_name = constituent_names(var_index), &
+             units = 'kg kg-1', &
+             vertical_dim = 'vertical_layer_dimension', &
+             min_value = 0.0_kind_phys, &
+             advected = .true., &
+             water_species = .true., &
+             mixing_ratio_type = 'wet', &
+             errcode = errcode, &
+             errmsg = errmsg)
+       else
```

Comment:
I believe we might need an `else if` here that checks if the constituent name also matches any of the names in `const_std_names` as well (which may need to be renamed to something like `water_species_std_names`), and if found will need the `water_species` and `mixing_ratio_type` flags set as well.  In other words:

```suggestion
       else if (any(const_std_names == trim(constituent_names(var_index)))) then
          call constituents(var_index)%instantiate( &
             std_name = constituent_names(var_index), &
             long_name = constituent_names(var_index), &
             units = 'kg kg-1', &
             vertical_dim = 'vertical_layer_dimension', &
             min_value = 0.0_kind_phys, &
             advected = .true., &
             water_species = .true., &
             mixing_ratio_type = 'wet', &
             errcode = errcode, &
             errmsg = errmsg)
       else
```

Reply:
a very good point - added `else if`